archive_converter.py

```diff
@@ -13,8 +13,69 @@
 # limitations under the License.
 
 import email
+import os
+from typing import List, Dict
 from setup_gmail import Message
 
+class ArchiveMessageIndex(object):
+    def __init__(self):
+        self._key_to_thread_start : Dict[str, Message] = {} #Maps message subject (str) to starting message of email thread
+        self._messages_seen : Dict[str, Message] = {} # Maps message.id to message
+ 
+    def find_thread(self, key: str) -> Message:
+        return self._key_to_thread_start[key]
+
+    def update(self, data_dir: str):
+        fully_updated = False
+        new_messages : List[Message] = []
+
+        for filename in os.listdir(data_dir):
+            if not filename.endswith(".txt"):
+                continue
+            email = generate_email_from_file(os.path.join(data_dir, filename))
+            if email.id not in self._messages_seen:
+                new_messages.append(email)
+        self._add_messages_to_index(new_messages)
+        
+    def size(self):
+        return len(self._messages_seen)
+
+    def _add_messages_to_index(self, new_messages : List[Message]):
+        ''' Iterates through all new emails and links together emails that form a thread by populating message.children. 
+        
+            ARGS:
+            new_messages::List[Message] - a list of messages that have yet to be added to the index
+        '''
+        need_parent : List[Message] = []
+            
+        # First iterates through all messages to distinguish between 1.) replies and 2.) the start of a thread.
+        for message in new_messages:
+            if message.id not in self._messages_seen:
+                self._messages_seen[message.id] = message
+            
+            # If message is a reply, it needs to be associated with an existing thread
+            if message.in_reply_to is not None:
+                need_parent.append(message)
+                continue
+                
+            # If message doesn't have the LKML style [PATCH v4 00/11],
+            # we don't need to put it in the index.
+            key = message.get_key()
+            if key:
+                self._key_to_thread_start[key] = message
+                
+        for message in need_parent:
+            if message.in_reply_to not in self._messages_seen:
+                print("No Parent Message for " + message.subject)
```

Comment:
nit: let's make this print be more clear what the implications are.
```python
'Could not find parent email, dropping ' + message.subject
```

Might be value in showing the offending `in_reply_to`, but this would be fine for now, imo